### PR TITLE
[ENH] Ensure that headers are consistent along the process

### LINKED ISSUE #710: CompCor contours misaligned with reference BOLD image
Unknown as yet whether this is just a visualization issue or if the misalignment shows bad masks that infect results.

DS000001 / sub-02

acompcor_misaligned

### PR DESCRIPTION
**Pull request by oesteban**
This PR fixes #710.

The problem was the corruption of xform matrices (sform, qform). Requires poldracklab/niworkflows#202

### CODE REVIEW

File: fmriprep/workflows/util.py [hunk 1]
```diff
@@ -81,10 +82,13 @@ def init_enhance_and_skullstrip_bold_wf(name='enhance_and_skullstrip_bold_wf',
                             name='combine_masks')
     apply_mask = pe.Node(fsl.ApplyMask(),
                          name='apply_mask')
+    copy_xform = pe.Node(CopyXForm(), name='copy_xform',
+                         mem_gb=0.1, run_without_submitting=True)
```
[the remark this answers is not in the document]
Comment (oesteban) [reply fmriprep/workflows/util.py]: That surprised me the first time. If you check output of this workflow, you'll see that it actually changes xforms. A suspicious 'descrip' field of 'FSL 5.0' makes me think that bet, apply mask or image maths may also change xforms. I don't see a problem on explicitly making sure that xforms are OK with this node.

This change of xform was the origin of #710.

[the remark this answers is not in the document]
Comment (effigies) [reply fmriprep/workflows/util.py]: That's maddening. I'm going to need to get mad about this again in the morning. Sorry if I've been asking dumb questions.

[the remark this answers is not in the document]
Comment (effigies) [reply fmriprep/workflows/util.py]: It's actually unifize that seems to be causing issues here:

```
$ fslhd skullstrip_first_pass/ref_image_corrected_brain.nii.gz | grep -A 14 qform_name 
qform_name     Scanner Anat
qform_code     1
qto_xyz:1      -3.125000  0.000000  -0.000000  97.931519
qto_xyz:2      0.000000  3.125000  -0.000000  -68.019211
qto_xyz:3      0.000000  0.000000  4.000000  -106.650566
qto_xyz:4      0.000000  0.000000  0.000000  1.000000
qform_xorient  Right-to-Left
qform_yorient  Posterior-to-Anterior
qform_zorient  Inferior-to-Superior
sform_name     Aligned Anat
sform_code     2
sto_xyz:1      -3.125000  0.000000  0.000000  98.909752
sto_xyz:2      0.000000  3.125000  0.000000  -110.716141
sto_xyz:3      0.000000  0.000000  4.000000  -51.800713
sto_xyz:4      0.000000  0.000000  0.000000  1.000000
```

```
$ fslhd unifize/uni.nii.gz | grep -A 14 qform_name 
qform_name     Talairach
qform_code     3
qto_xyz:1      -3.125000  -0.000000  -0.000000  98.909752
qto_xyz:2      0.000000  3.125000  -0.000000  -110.716141
qto_xyz:3      -0.000000  0.000000  4.000000  -51.800713
qto_xyz:4      0.000000  0.000000  0.000000  1.000000
qform_xorient  Right-to-Left
qform_yorient  Posterior-to-Anterior
qform_zorient  Inferior-to-Superior
sform_name     Talairach
sform_code     3
sto_xyz:1      -3.125000  -0.000000  -0.000000  98.909752
sto_xyz:2      -0.000000  3.125000  -0.000000  -110.716141
sto_xyz:3      0.000000  0.000000  4.000000  -51.800713
sto_xyz:4      0.000000  0.000000  0.000000  1.000000
```

... What?

[the remark this answers is not in the document]
Comment (effigies) [reply fmriprep/workflows/util.py]: I've submitted an issue (afni/afni#65), but unless neurodebian starts updating AFNI regularly (or we start using a more  recent neurodocker build), we're still going to need to work around this for the foreseeable future.

[the remark this answers is not in the document]
Comment (oesteban) [reply fmriprep/workflows/util.py]: Great job! For the 'descrip' header I assumed it was FSL, but that wouldn't rule out unifize.

**Comment by oesteban** [comment]
I'm double checking this on the dataset with freesurfer on. Also checking if this fixes also #708

**Comment by oesteban** [comment]
Bad news: this PR either does not fix #710 when freesurfer is used or I'm seeing #703 again for reusing the pre-computed freesurfer. Either way I need to get to the bottom before merging this.

File: fmriprep/workflows/util.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (effigies) [reply fmriprep/workflows/util.py]: Okay, it looks like this is a [WONTFIX](https://github.com/afni/afni/issues/65#issuecomment-335496357). We'll want to consider how we interface with AFNI tools in general.

Thinking more long-term, I've proposed nipy/nipype#2223.

**Comment by oesteban** [comment]
So, with the new code I don't see the problem of #708 on ds101, sub-06. I'll check more subjects from that PR.

On the other hand, I repeated (reusing pre-computed freesurfer) on ds001, sub-02 and this issue is still happening.

**Comment by oesteban** [comment]
I am considering that we are now dealing with a side effect of #741 for two reasons:

- I've checked the headers of the inputs (the reference BOLD for bbregister and the split volumes for the resampling step) and now they match correctly. This was wrong before the changes in this PR.
- The BBR path seems to work well (as a consequence of the previous fix) so we are probable looking at a bbregister problem and the conversion of matrices to ITK.

WDYT @effigies?

**Comment by effigies** [comment]
Ugh. That's possible. (Though at least on my tests, I get the exact same results whether we go through your branch or 0.6.5.)

**Comment by effigies** [comment]
Oh, I think I see the issue. The bbregister-targeted file is T1w, but the FLIRT-targeted file is T1w downsampled to BOLD resolution. There's been no update step targeting the BOLD-resolution T1w file.

**Comment by oesteban** [comment]
The resulting ITK transform should be in mm (meaning, the resolution of the target space is not relevant). I'm going to check apply the transform using the full-resolution T1w as reference and I bet we will get the same.

**Comment by oesteban** [comment]
Basically, what the affine transform does is: for each node of the sampling grid, get the coordinates in mm (physical space) and find the coordinates in mm that correspond to that point in moving space.

Therefore, changing the fixed image resolution you only change the location of the points you want to find in moving space. The transform is the same for both images (if it is defined in mm like ITK, this changes if you define transforms in pixels like legacy registration algorithms).

**Comment by oesteban** [comment]
The `out_type='RAS2RAS'` did not make any difference :(

**Comment by effigies** [comment]
Okay, if I manually `lta_convert` to FSL and then `c3d_affine_tool` to ITK, then I get alignment. Want me to send you a patch, or should we keep going until we find something more elegant?

**Comment by oesteban** [comment]
I'm fine with that patch, but you put a lot of work on those LTA, so I'm
not going to rush for a fix. It's your call :)

**Comment by effigies** [comment]
I don't love it, but I added a hacky solution in oesteban#2. I've verified that it fixes alignment for me.

I'm going to ping FreeSurfer to see if `lta_convert` can get fixed by 6.0.1.

Also, so we have it written down somewhere, `c3d_affine_tool` can do both fwd and inv transforms in a single shot. I just didn't want to go through another round of nipype editing to get this in.

```Shell
c3d_affine_tool -ref <ref> -src <src> in_xfm.mat -fsl2ras -oitk itk.txt -inv -oitk itk_inv.txt
```

Turns out order of operations matters a whole lot, though, which also makes designing an update to the interface a little weird. I guess you could add an `inverse_itk_transform = File(argstr='-inv -oitk %s', position=6)`.

**Comment by oesteban** [comment]
Just not that the fix worked well on ds001, sub-02. I'll test more subjects from #708

**Comment by oesteban** [comment]
Bummer, I just noticed this error in the end of the report https://320-53608443-gh.circle-artifacts.com/0/home/ubuntu/ds054/out/fmriprep/sub-100185.html

**Pull request merged**